mesonbuild/modules/python.py

```diff
@@ -159,32 +161,90 @@ def extension_module_method(self, args: T.Tuple[str, T.List[BuildTargetSource]],
 
             kwargs['install_dir'] = self._get_install_dir_impl(False, subdir)
 
+        target_suffix = self.suffix
+
         new_deps = mesonlib.extract_as_list(kwargs, 'dependencies')
-        has_pydep = any(isinstance(dep, _PythonDependencyBase) for dep in new_deps)
-        if not has_pydep:
+        pydep = next((dep for dep in new_deps if isinstance(dep, _PythonDependencyBase)), None)
+        if pydep is None:
             pydep = self._dependency_method_impl({})
             if not pydep.found():
                 raise mesonlib.MesonException('Python dependency not found')
             new_deps.append(pydep)
             FeatureNew.single_use('python_installation.extension_module with implicit dependency on python',
                                   '0.63.0', self.subproject, 'use python_installation.dependency()',
                                   self.current_node)
+
+        limited_api_version = kwargs.pop('limited_api')
+        if limited_api_version != '':
+
+            target_suffix = self.limited_api_suffix
+
+            limited_api_version_hex = self._convert_api_version_to_py_version_hex(limited_api_version)
+            if mesonlib.version_compare(limited_api_version, '<3.2'):
+                raise InvalidArguments(f'Python Limited API version invalid: {limited_api_version} (must be greater than 3.2)')
+            if mesonlib.version_compare(limited_api_version, '>' + pydep.version):
+                raise InvalidArguments(f'Python Limited API version too high: {limited_api_version} (detected {pydep.version})')
+
+            new_c_args = mesonlib.extract_as_list(kwargs, 'c_args')
+            new_c_args.append(f'-DPy_LIMITED_API={limited_api_version_hex}')
+            kwargs['c_args'] = new_c_args
+
+            # When compiled under MSVC, Python's PC/pyconfig.h forcibly inserts pythonMAJOR.MINOR.lib
+            # into the linker path when not running in debug mode via a series #pragma comment(lib, "")
+            # directives. We manually override these here as this interferes with the intended
+            # use of the 'limited_api' kwarg
```

Comment:
It is a shame you need this. I think the intention of [the header](https://github.com/python/cpython/blob/a73daf54ebd7bd6bf32e82766a605ebead2f128c/PC/pyconfig.h#L312) was deal with linking the correct import lib, and meson should ideally not specify any link arguments to deal with the import lib at all. But it seems the header does not properly link to the "limited api + debug" import lib. I opened https://github.com/python/cpython/issues/107585 to clarify that

Reply:
I also regret that this was necessary but it does seem that CPython's `pyconfig.h` is missing a bit of logic here.

Back when I was starting this work I recall finding a PR in CPython which aimed to fix the logic in `PC/pyconfig.h`, allowing a correct and peaceful coexistence between the `Py_LIMITED_API` and `_DEBUG` macros, but after searching for it again just now I can't seem to find it. That's frustrating: if I'm unable to find it again then I will work on a fix for CPython.

That being said, even if a fix makes it into CPython, I think Meson should probably still aim to support users with older versions of CPython that are already in the wild, right?

Reply:
Precisely.